MonoGame.Framework/Utilities/OggStream.cs

```diff
@@ -0,0 +1,504 @@
+        }
+
+        public void Dispose()
+        {
+            var state = AL.GetSourceState(alSourceId);
+            if (state == ALSourceState.Playing || state == ALSourceState.Paused)
+                StopPlayback();
+
+            lock (prepareMutex)
+            {
+                OggStreamer.Instance.RemoveStream(this);
+
+                if (state != ALSourceState.Initial)
+                    Empty();
+
+                Close();
+
+                underlyingStream.Dispose();
+            }
+
+            AL.DeleteSource(alSourceId);
+            AL.DeleteBuffers(alBufferIds);
+
+            if (ALHelper.Efx.IsInitialized)
+                ALHelper.Efx.DeleteFilter(alFilterId);
+
+            ALHelper.Check();
+        }
+
+        void StopPlayback()
+        {
+            AL.SourceStop(alSourceId);
+            ALHelper.Check();
+        }
+
+        void Empty()
+        {
+            int queued;
+            AL.GetSource(alSourceId, ALGetSourcei.BuffersQueued, out queued);
+            if (queued > 0)
+            {
+                try
+                {
+                    AL.SourceUnqueueBuffers(alSourceId, queued);
+                    ALHelper.Check();
+                }
+                catch (InvalidOperationException)
+                {
+                    // This is a bug in the OpenAL implementation
+                    // Salvage what we can
+                    int processed;
+                    AL.GetSource(alSourceId, ALGetSourcei.BuffersProcessed, out processed);
+                    var salvaged = new int[processed];
+                    if (processed > 0)
+                    {
+                        AL.SourceUnqueueBuffers(alSourceId, processed, salvaged);
+                        ALHelper.Check();
+                    }
+
+                    // Try turning it off again?
+                    AL.SourceStop(alSourceId);
+                    ALHelper.Check();
+
+                    Empty();
+                }
+            }
+        }
+
+        internal void Open(bool precache = false)
+        {
+            underlyingStream.Seek(0, SeekOrigin.Begin);
+            Reader = new VorbisReader(underlyingStream, false);
+
+            if (precache)
+            {
+                // Fill first buffer synchronously
+                OggStreamer.Instance.FillBuffer(this, alBufferIds[0]);
+                AL.SourceQueueBuffer(alSourceId, alBufferIds[0]);
+                ALHelper.Check();
+
+                // Schedule the others asynchronously
+                OggStreamer.Instance.AddStream(this);
+            }
+
+            Ready = true;
+        }
+
+        internal void Close()
+        {
+            if (Reader != null)
+            {
+                Reader.Dispose();
+                Reader = null;
+            }
+            Ready = false;
+        }
+    }
+
+    public class OggStreamer : IDisposable
+    {
+        Action finishedPlaying = null;
```

Comment:
coding note 2, I was going to put List<Action> at the end so that multiple song instances would call the corresponding correct methods.